### PR TITLE
adding flash attention and xformer memory efficient through PT SDPA

### PR DESCRIPTION
**Pull request by HamidShojanazeri**
# What does this PR do?

This PR adds the Flash Attention and Xformer mem-efficient kernel through PT SDPA, this work has been integrated with `optimum` library of HF, read more about [here](https://pytorch.org/blog/out-of-the-box-acceleration/).

Tested on 7B for FSDP only had a nice 30% speed up, for FSDP+PEFT 5% and not much on PEFT+quantization/1 gpu.

Fixes # (issue)
No related issue.

## Feature/Issue validation/testing

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [ ] Test A : Logs/ perf number of with out this feature with 10 steps : avg epoch time 55.17s
[Logs for Test A
](https://gist.github.com/HamidShojanazeri/e6cc9eb3ca9a3bf5a83174c8eb2c5a18)

- [ ] Test B : Logs/ perf number of with this feature with 10 steps : avg epoch time 42.44
[Logs for Test B](https://gist.github.com/HamidShojanazeri/587755cebc0aa0e67edf0995a70f8127)

## Before submitting
- [x ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue? Please add a link
      to it if that's the case.
- [x ] Did you make sure to update the documentation with your changes?  
- [ ] Did you write any new necessary tests?

Thanks for contributing 🎉!

### CODE REVIEW

File: inference/chat_completion.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 
 from peft import PeftModel, PeftConfig
 from transformers import LlamaConfig, LlamaTokenizer, LlamaForCausalLM
+from optimum.bettertransformer import BetterTransformer
```
Comment (chauhang) [inline inference/chat_completion.py:14]: Please remove from here as we have this in the try-catch-block for optional install

Comment (HamidShojanazeri) [reply inference/chat_completion.py]: Done

File: inference/chat_completion.py [hunk 2]
```diff
@@ -57,6 +59,18 @@ def main(
     torch.cuda.manual_seed(seed)
     torch.manual_seed(seed)
     model = load_model(model_name, quantization)
+    if use_fast_kernels:
```
Comment (chauhang) [inline inference/chat_completion.py:61]: For the PEFT model case, does this need to happen after the PEFT/Lora weights have been merged?

File: llama_finetuning.py [hunk 3]
```diff
@@ -24,6 +24,7 @@
     BitsAndBytesConfig
 )
 import torch.distributed as dist
+from optimum.bettertransformer import BetterTransformer
```
Comment (chauhang) [inline llama_finetuning.py:27]: Please remove from here as we have this in the try-catch-block for optional install of optimum library

**Review by chauhang — CHANGES_REQUESTED** [review]
@HamidShojanazeri Thanks for adding the BT optimizations. Please see the comments inline

File: inference/chat_completion.py (hunk 2 above)
Comment (HamidShojanazeri) [reply inference/chat_completion.py:63]: sure, updated works both way.

File: llama_finetuning.py [hunk 4]
```diff
@@ -95,7 +94,17 @@ def main(**kwargs):
...
+        """
+        For FSDP and FSDP+PEFT, setting 'use_fast_kernels' will enable
+        using of Flash Attention or Xformer memory-efficient kernels 
+        based on the hardware being used. This would speed up fine-tuning.
+        """
+        try:
+            from optimum.bettertransformer import BetterTransformer
+        except ImportError:
+            print("Module 'optimum' not found. Please install 'optimum' it before proceeding.")
+        model = BetterTransformer.transform(model)
```
Comment (chauhang) [inline llama_finetuning.py:107]: This needs to be inside the try catch for handling the failure condition

File: inference/inference.py [hunk 5]
```diff
@@ -51,6 +52,23 @@ def main(
...
+        Setting 'use_fast_kernels' will enable
+        using of Flash Attention or Xformer memory-efficient kernels 
+        based on the hardware being used. This would speed up inference when used for batched inputs.
+        """
+        try:
+            from optimum.bettertransformer import BetterTransformer
+        except ImportError:
+            print("Module 'optimum' not found. Please install 'optimum' it before proceeding.")
+
+        model = BetterTransformer.transform(model)
```
Comment (chauhang) [inline inference/inference.py:71]: This needs to be inside the try catch for handling the failure condition

File: inference/chat_completion.py [hunk 6]
```diff
@@ -59,6 +60,19 @@ def main(
...
+        Setting 'use_fast_kernels' will enable
+        using of Flash Attention or Xformer memory-efficient kernels 
+        based on the hardware being used. This would speed up inference when used for batched inputs.
+        """
+        try:
+            from optimum.bettertransformer import BetterTransformer
+        except ImportError:
+            print("Module 'optimum' not found. Please install 'optimum' it before proceeding.")
+
+        model = BetterTransformer.transform(model)
```
Comment (chauhang) [inline inference/chat_completion.py:74]: This needs to be inside the try catch for handling the failure condition

**Review by chauhang — CHANGES_REQUESTED** [review]
Please fix the issues and also attach the logs for the inference speedup.

**Review by chauhang — APPROVED** [review]

**Pull request merged**